dll/win32/shell32/CCopyAsPathMenu.cpp

```diff
@@ -0,0 +1,228 @@
+    CDataObjectHIDA pCIDA(pdto);
+    HRESULT hr = pCIDA.hr();
+    if (SUCCEEDED(hr))
+    {
+        for (i = 0, count = pCIDA->cidl; i < count && SUCCEEDED(hr); ++i)
+        {
+            PCUIDLIST_ABSOLUTE folder = HIDA_GetPIDLFolder(pCIDA);
+            PCUIDLIST_RELATIVE item = HIDA_GetPIDLItem(pCIDA, i);
+            CComHeapPtr<ITEMIDLIST> full;
+            hr = SHILCombine(folder, item, &full);
+            if (SUCCEEDED(hr))
+            {
+                PCUITEMID_CHILD child;
+                CComPtr<IShellFolder> sf;
+                hr = SHBindToParent(full, IID_PPV_ARG(IShellFolder, &sf), &child);
+                if (SUCCEEDED(hr))
+                {
+                    STRRET strret;
+                    hr = sf->GetDisplayNameOf(child, SHGDN_FORPARSING, &strret);
+                    if (SUCCEEDED(hr))
+                    {
+                        CComHeapPtr<WCHAR> path;
+                        hr = StrRetToStrW(&strret, child, &path);
+                        if (SUCCEEDED(hr))
+                        {
+                            AppendToPathList(paths, path, i);
+                        }
+                    }
+                }
+            }
+        }
+    }
+    else
+    {
+        FORMATETC fmte = { CF_HDROP, NULL, DVASPECT_CONTENT, -1, TYMED_HGLOBAL };
+        STGMEDIUM stgm;
+        hr = pdto->GetData(&fmte, &stgm);
+        if (SUCCEEDED(hr))
+        {
+            for (i = 0, count = DragQueryFileW((HDROP)stgm.hGlobal, -1, NULL, 0); i < count && SUCCEEDED(hr); ++i)
+            {
+                WCHAR path[MAX_PATH];
+                if (DragQueryFileW((HDROP)stgm.hGlobal, i, path, _countof(path)))
+                {
+                    AppendToPathList(paths, path, i);
+                }
+            }
+            ReleaseStgMedium(&stgm);
+        }
+    }
+#endif
+
+    if (SUCCEEDED(hr))
+    {
+        DWORD err = SetClipboardFromString(paths);
+        hr = HRESULT_FROM_WIN32(err);
+    }
+
+    return hr;
+}
+
+STDMETHODIMP
+CCopyAsPathMenu::QueryContextMenu(HMENU hMenu, UINT indexMenu, UINT idCmdFirst, UINT idCmdLast, UINT uFlags)
+{
+    MENUITEMINFOW mii;
+    UINT Count = 0;
+
+    TRACE("CCopyAsPathMenu::QueryContextMenu(%p %p %u %u %u %u)\n", this,
+          hMenu, indexMenu, idCmdFirst, idCmdLast, uFlags);
+
+    // Detect SHIFT key press
+    if (!(GetKeyState(VK_SHIFT) & 0x8000))
+        return MAKE_HRESULT(SEVERITY_SUCCESS, 0, idCmdFirst);
+
+    if (uFlags & (CMF_NOVERBS | CMF_VERBSONLY))
+        return MAKE_HRESULT(SEVERITY_SUCCESS, 0, idCmdFirst);
+
+    m_idCmdFirst = idCmdFirst;
+
+    // Insert "Copy as path"
+    CStringW strText(MAKEINTRESOURCEW(IDS_COPYASPATHMENU));
+    ZeroMemory(&mii, sizeof(mii));
+    mii.cbSize = sizeof(mii);
+    mii.fMask = MIIM_ID | MIIM_TYPE;
+    mii.fType = MFT_STRING;
+    mii.wID = m_idCmdFirst;
+    mii.dwTypeData = strText.GetBuffer();
+    mii.cch = wcslen(mii.dwTypeData);
```

Comment:
I don't think you need to set the length when "writing". Only when "reading" menu items.

Reply:
Removed.